solidity/contracts/CheckBitcoinSigs.sol

```diff
@@ -19,33 +17,39 @@ library CheckBitcoinSigs {
     function accountFromPubkey(bytes memory _pubkey) internal pure returns (address) {
         require(_pubkey.length == 64, "Pubkey must be 64-byte raw, uncompressed key.");
 
-        // keccak hash of uncompressed unprefixed pubkey
-        bytes32 _digest = keccak256(_pubkey);
+        bytes32 _digest = _pubkey.ref(0).keccak();
         return address(uint256(_digest));
     }
 
     /// @notice          Calculates the p2wpkh output script of a pubkey
-    /// @dev             Compresses keys to 33 bytes as required by Bitcoin
+    /// @dev             Compresses keys to 33 bytes as required by Bitcoin wpkh
     /// @param _pubkey   The public key, compressed or uncompressed
     /// @return          The p2wkph output script
-    function p2wpkhFromPubkey(bytes memory _pubkey) internal pure returns (bytes memory) {
-        bytes memory _compressedPubkey;
-        uint8 _prefix;
-
-        if (_pubkey.length == 64) {
-            _prefix = uint8(_pubkey[_pubkey.length - 1]) % 2 == 1 ? 3 : 2;
-            _compressedPubkey = abi.encodePacked(_prefix, _pubkey.slice(0, 32));
-        } else if (_pubkey.length == 65) {
-            _prefix = uint8(_pubkey[_pubkey.length - 1]) % 2 == 1 ? 3 : 2;
-            _compressedPubkey = abi.encodePacked(_prefix, _pubkey.slice(1, 32));
-        } else {
-            _compressedPubkey = _pubkey;
+    function p2wpkhFromPubkey(bytes memory _pubkey) internal view returns (bytes memory) {
+        bytes29 _pubkeyView = _pubkey.ref(0);
+
+        bytes20 _digest;
+        if (_pubkey.length == 33) {_digest = _pubkeyView.hash160();}
+        else if (_pubkey.length == 64) {
+            // This is less memory efficient than it ought to be
+            // But we immediately deallocate it, so it's usually no issue
+            uint8 prefix = uint8(_pubkey[_pubkey.length - 1]) % 2 == 1 ? 3 : 2;
+            bytes memory _x = _pubkey.ref(0).prefix(32, 0).clone();
+            bytes memory _compressed = abi.encodePacked(prefix, _x);
+            _digest = _compressed.ref(0).hash160();
```

Comment:
This seems like it could be simplified... Maybe:
```
uint8 prefix = uint8(_pubkey[pubkey.length - 1]) % 2 + 2;
bytes memory _x = _pubkeyView.prefix(32, 0);
```
Also, does _x need to be saved to memory?  Could we do something like:
```
bytes memory _compressed =  abi.encodePacked(prefix, _pubkey.safeSlice(0, 32)); // not sure if there is a safeSlice equivalent in Solidity...
_digest = compressed.ref(0).hash160();
```

Reply:
> This seems like it could be simplified... Maybe:

Good call

> Also, does _x need to be saved to memory 

I chose to do it this way because of some rough edges on solidity memory. Because prefix isn't in memory, we can't use `unsafeJoin`. We could drop into assembly and store the prefix, then x. but I'd prefer to avoid doing that if it can be avoided. This uses 32 bytes more memory than I'd prefer, but I think it's okay